File: backend/lan_playground/domain/commands.py
"""Command envelope per docs/INFINITE_STACKS_CONTRACTS.md §2."""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class CommandType(str, Enum):
    JOIN_RUN = "join_run"
    MOVE = "move"
    BREACH = "breach"
    OBSERVE = "observe"
    INSPECT = "inspect"
    PASS = "pass"
    CHECK = "check"
    INSPECT_OBJECT = "inspect_object"
    SUBMIT_SOLUTION = "submit_solution"
    REQUEST_HINT = "request_hint"
    COMBAT_ATTACK = "combat_attack"
    COMBAT_MANEUVER = "combat_maneuver"
    COMBAT_REACTION = "combat_reaction"
    COMBAT_MOVE = "combat_move"
    COMBAT_QUICK_INTERACTION = "combat_quick_interaction"
    COMBAT_STABILIZE = "combat_stabilize"
    COMBAT_BARRICADE = "combat_barricade"
    COMBAT_END_TURN = "combat_end_turn"

    # Wave-5 enemy-to-hit / reaction-interrupt window (board task #16,
    # stacks-enemyroll; domain registration requested via chat 2026-07-19,
    # landed by stacks-shopwire as domain schema owner)
    RESOLVE_REACTION = "resolve_reaction"

    # Heroes wiring (infinite_stacks.md §11, §13; wave 4, board task #13)
    ROLL_ATTRIBUTE_DICE = "roll_attribute_dice"
    CREATE_HERO = "create_hero"
    PLAY_CARD = "play_card"
    DRAW_CARDS = "draw_cards"
    SAFE_REST = "safe_rest"
    PICKUP_ITEM = "pickup_item"
    DROP_ITEM = "drop_item"
    TRADE_ITEM = "trade_item"
    RECOVER_BODY_LOOT = "recover_body_loot"

    # Shops wiring (infinite_stacks.md §9.6, §16.4-16.6; wave 5, board task #18)
    SHOP_BUY = "shop_buy"
    SHOP_SELL = "shop_sell"
    SHOP_REPAIR = "shop_repair"
    SHOP_IDENTIFY = "shop_identify"
    SHOP_TREAT = "shop_treat"

    # Server-side clue sharing (wave 5, board task #18)
    SHARE_CLUE = "share_clue"

    # Abilities (wave 6, board task #21; docs/PLAYTEST_FINDINGS_2026-07-19.md
    # E1/A5) -- player-invoked trigger=="manual" abilities. Domain schema
    # posted to the collab room 2026-07-20.
    USE_ABILITY = "use_ability"

    # Study-room domain wiring (wave6b/slice-wiring, docs/
    # INFINITE_STACKS_STUDY_SLICE.md, wavebasedgame.md §3.2-3.6): object/NPC
    # interaction commands realizing the revised core loop. `interact`
    # payload: {object_id: str, interaction_id: str}. `converse` payload:
    # {npc_id: str, appeal_objective_id: str | None} -- NEITHER modifier
    # input is ever taken from the payload as a tier (standing rule #5: no
    # client-supplied modifiers, ever). Evidence is derived server-side from
    # whether `systems.study_social_wire.EVIDENCE_FACT_ID` has already been
    # promoted for the acting hero's own viewer id (StudyRoomState.
    # promoted_fact_ids). Motive is derived server-side from the optional
    # `appeal_objective_id` (a roleplay choice naming which NPC objective the
    # hero appeals to) matched against the NPC's own authored objective data
    # -- see systems/study_social_wire.py::_derive_motive_alignment. An
    # unknown/garbage appeal id degrades to NEUTRAL, never an error. Any
    # legacy `motive_alignment` payload field is ignored entirely.
    INTERACT = "interact"
    CONVERSE = "converse"


class ErrorCode(str, Enum):
    STALE_REVISION = "stale_revision"
    ILLEGAL_ACTION = "illegal_action"
    NOT_YOUR_TURN = "not_your_turn"
    UNKNOWN_TARGET = "unknown_target"
    SCHEMA_ERROR = "schema_error"


@dataclass(frozen=True)
class Command:
    command_id: str
    idempotency_key: str
    run_id: str
    type: CommandType
    hero_id: str | None = None
    encounter_id: str | None = None
    expected_revision: int = 0
    payload: dict = field(default_factory=dict)


class CommandError(Exception):
    def __init__(self, code: ErrorCode, message: str, legal_actions: list[str] | None = None):
        super().__init__(message)
        self.code = code
        self.message = message
        self.legal_actions = legal_actions or []

    def to_dict(self) -> dict:
        return {"code": self.code.value, "message": self.message, "legal_actions": self.legal_actions}

File: backend/lan_playground/static/src/core/commands.js
// Builds command envelopes matching docs/INFINITE_STACKS_CONTRACTS.md S2.
// Pure functions only: no network, no DOM, no store access. Callers
// (screens/map.js via main.js) supply expectedRevision from the store and
// pass the resulting envelope to core/socket.js's sendCommand.

function generateId() {
  if (window.crypto && typeof window.crypto.randomUUID === "function") {
    return window.crypto.randomUUID();
  }
  return `id-${Date.now()}-${Math.random().toString(16).slice(2)}`;
}

export function buildCommand(type, payload, { expectedRevision, encounterId = null, idempotencyKey } = {}) {
  return {
    command_id: generateId(),
    idempotency_key: idempotencyKey || generateId(),
    encounter_id: encounterId,
    expected_revision: expectedRevision,
    type,
    payload: payload || {},
  };
}

export function moveCommand(toRoomId, expectedRevision) {
  return buildCommand("move", { to_room_id: toRoomId }, { expectedRevision });
}

export function breachCommand(direction, expectedRevision) {
  return buildCommand("breach", { direction }, { expectedRevision });
}

export function observeCommand(direction, expectedRevision) {
  return buildCommand("observe", { direction }, { expectedRevision });
}

export function inspectCommand(expectedRevision) {
  return buildCommand("inspect", {}, { expectedRevision });
}

export function passCommand(expectedRevision) {
  return buildCommand("pass", {}, { expectedRevision });
}

export function checkCommand(dc, expectedRevision) {
  return buildCommand("check", { dc }, { expectedRevision });
}

// Wave-2 puzzle commands (docs/INFINITE_STACKS_CONTRACTS.md S2 "Wave-2
// additions"). Payload keys/types match the domain vocabulary exactly:
// inspect_object takes object_id, submit_solution takes an ordered list of
// item ids under the key "solution" (never "answer"), request_hint takes no
// payload at all -- requesting a hint past the last available one is how the
// server force-progresses the room (systems/puzzles.py's handle_request_hint),
// there is no separate force_progress command.
export function inspectObjectCommand(objectId, expectedRevision) {
  return buildCommand("inspect_object", { object_id: objectId }, { expectedRevision });
}

export function submitSolutionCommand(solution, expectedRevision) {
  return buildCommand("submit_solution", { solution }, { expectedRevision });
}

export function requestHintCommand(expectedRevision) {
  return buildCommand("request_hint", {}, { expectedRevision });
}

// Wave-3 combat commands, per stacks-conflict's 17:15 vocabulary post
// (docs/INFINITE_STACKS_COMBAT.md's package + the director's ruling that
// removed client-supplied numeric modifiers): combat_attack/combat_maneuver
// take only target_id/attribute/skill(/maneuver) -- no accuracy or damage
// numbers from the client, those are always server-resolved. EARLY DRAFT:
// stacks-conflict said to expect at most small renames once their code lands.
export function combatAttackCommand(targetId, attribute, skill, expectedRevision, encounterId) {
  return buildCommand("combat_attack", { target_id: targetId, attribute, skill }, { expectedRevision, encounterId });
}

export function combatManeuverCommand(maneuver, targetId, attribute, skill, expectedRevision, encounterId) {
  return buildCommand("combat_maneuver", { maneuver, target_id: targetId, attribute, skill }, { expectedRevision, encounterId });
}

// §21.4 reaction interrupt window (stacks-enemyroll's wave-5 transport-
// injection spec, board task #16): resolve_reaction takes only the pending
// interrupt's reaction_id (so a stale/duplicate click can never resolve the
// wrong reaction) plus the chosen reaction name -- no client-supplied
// incoming_attack_total/incoming_damage/position, unlike the wave-3
// freeform combat_reaction command it replaces in the UI (that one required
// exactly the raw numeric modifiers this codebase's wire contract forbids
// from ever leaving the client; see combat.js's renderReactionPrompt).
export function resolveReactionCommand(reactionId, reaction, expectedRevision, encounterId) {
  return buildCommand("resolve_reaction", { reaction_id: reactionId, reaction }, { expectedRevision, encounterId });
}

export function combatEndTurnCommand(expectedRevision, encounterId) {
  return buildCommand("combat_end_turn", {}, { expectedRevision, encounterId });
}

// Wave-5 hero commands (docs/INFINITE_STACKS_CONTRACTS.md S5.4, board task
// #17): payload keys/types match the domain vocabulary systems/heroes_wire.py
// validates exactly. No raw numeric modifier ever leaves the client here --
// attribute_assignment sends the die VALUES the player rolled (server-
// supplied, already public via attribute_dice_rolled), never a made-up
// score, and item/card selections are always ids.
export function rollAttributeDiceCommand(expectedRevision) {
  return buildCommand("roll_attribute_dice", {}, { expectedRevision });
}

export function createHeroCommand(
  { name, backgroundId, attributeAssignment, generalCardIds, personaCardId, equipmentCardIds, avatarId, color },
  expectedRevision,
) {
  return buildCommand(
    "create_hero",
    {
      name,
      background_id: backgroundId,
      attribute_assignment: attributeAssignment,
      general_card_ids: generalCardIds,
      persona_card_id: personaCardId,
      equipment_card_ids: equipmentCardIds || [],
      // F1 tokens (playtest): CONFIRMED contract, stacks-abilities 04:30 --
      // validated + auto-assigned server-side if omitted (null is legal).
      avatar_id: avatarId || null,
      color: color || null,
    },
    { expectedRevision },
  );
}

export function drawCardsCommand(count, expectedRevision) {
  return buildCommand("draw_cards", { count }, { expectedRevision });
}

export function playCardCommand(cardId, { targetHeroId = null, targetEnemyId = null } = {}, expectedRevision, encounterId) {
  return buildCommand(
    "play_card",
    { card_id: cardId, target_hero_id: targetHeroId, target_enemy_id: targetEnemyId },
    { expectedRevision, encounterId },
  );
}

export function safeRestCommand(expectedRevision) {
  return buildCommand("safe_rest", {}, { expectedRevision });
}

export function pickupItemCommand(itemInstanceId, expectedRevision) {
  return buildCommand("pickup_item", { item_instance_id: itemInstanceId }, { expectedRevision });
}

export function dropItemCommand(itemId, expectedRevision) {
  return buildCommand("drop_item", { item_id: itemId }, { expectedRevision });
}

export function tradeItemCommand(toHeroId, itemId, expectedRevision) {
  return buildCommand("trade_item", { to_hero_id: toHeroId, item_id: itemId }, { expectedRevision });
}

export function recoverBodyLootCommand(deadHeroId, itemIds, expectedRevision) {
  return buildCommand("recover_body_loot", { dead_hero_id: deadHeroId, item_ids: itemIds || null }, { expectedRevision });
}

// D1 abilities panel "Use" control: stacks-abilities' 04:30 post names the
// command `use_ability`, charge-gated the same way signature charges are.
// If the server hasn't wired it yet this resolves to an ordinary
// CommandError, handled the same as any other rejected command.
export function useAbilityCommand(abilityId, expectedRevision) {
  return buildCommand("use_ability", { ability_id: abilityId }, { expectedRevision });
}

// Wave-6B part 4 (docs/INFINITE_STACKS_CONTRACTS.md S5.11): interact/converse
// payloads match the domain vocabulary exactly. `interact` names an object +
// one of its own declared interaction ids -- never a free-text verb.
// `converse` never takes a modifier/tier from the client (standing rule #5):
// `appealObjectiveId` is a ROLEPLAY CHOICE naming which of the NPC's own
// disclosed objectives the hero is appealing to (or omitted entirely for "no
// appeal") -- the server derives the actual MotiveAlignment from its own
// authored NPC data, never trusting a client-supplied strength/tier. A
// legacy `motive_alignment` field is deliberately never built here at all.
export function interactCommand(objectId, interactionId, expectedRevision) {
  return buildCommand("interact", { object_id: objectId, interaction_id: interactionId }, { expectedRevision });
}

export function converseCommand(npcId, appealObjectiveId, expectedRevision) {
  return buildCommand("converse", { npc_id: npcId, appeal_objective_id: appealObjectiveId || null }, { expectedRevision });
}
